beacon/handlers.go

```diff
@@ -510,11 +502,45 @@ func (pd *ProtocolDriver) registerProposed(logger log.Log, epoch types.EpochID,
 	return pd.states[epoch].registerProposed(logger, nodeID)
 }
 
-func (pd *ProtocolDriver) registerVoted(logger log.Log, epoch types.EpochID, nodeID types.NodeID, round types.RoundID) error {
+func (pd *ProtocolDriver) registerVoted(epoch types.EpochID, nodeID types.NodeID, round types.RoundID) error {
 	pd.mu.Lock()
 	defer pd.mu.Unlock()
 	if _, ok := pd.states[epoch]; !ok {
 		return errEpochNotActive
 	}
-	return pd.states[epoch].registerVoted(logger, nodeID, round)
+	return pd.states[epoch].registerVoted(nodeID, round)
+}
+
+const wordSize = 64
+
+func newVotesTracker(limit uint32) *votesTracker {
+	b := bits.Len32(limit)
+	words := b / wordSize
+	if b%wordSize != 0 {
+		words += 1
+	}
+	return &votesTracker{words: make([]uint64, words)}
+}
+
+type votesTracker struct {
+	words []uint64
+}
+
+func (v *votesTracker) register(round types.RoundID) bool {
+	if uint32(round) > wordSize*uint32(len(v.words)) {
+		return false
+	}
+	word := uint64(round) / wordSize
+	position := uint64(1) << (uint64(round) % wordSize)
+	if v.words[word]&position > 0 {
+		return false
+	}
+	v.words[word] |= position
+	return true
+}
+
+func (v *votesTracker) voted(round types.RoundID) bool {
```

Comment:
it's strange that you check for `round > wordSize*uint32(len(v.words))` in `register` but not in `voted`.